java/arcs/core/util/Scheduler.kt

```diff
@@ -11,141 +11,135 @@
+import kotlinx.coroutines.channels.ConflatedBroadcastChannel
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.consumeAsFlow
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
 import kotlinx.coroutines.suspendCancellableCoroutine
 import kotlinx.coroutines.withTimeout
 
+@Suppress("UNUSED_PARAMETER")
+@Deprecated(
+    "Time is no longer of the essence",
+    ReplaceWith("Scheduler(context, agendaProcessingTimeoutMs)")
+)
+fun Scheduler(
+    time: Time,
+    context: CoroutineContext,
+    agendaProcessingTimeoutMs: Long = DEFAULT_AGENDA_PROCESSING_TIMEOUT_MS
+): Scheduler = Scheduler(context, agendaProcessingTimeoutMs)
+
 /**
  * The [Scheduler] is responsible for scheduling the execution of a batch of [Task]s (known as an
- * [Agenda]), at a defined maximum [scheduleRateHz], where each agenda is limited to running within
- * the specified [agendaProcessingTimeoutMs] window.
+ * [Agenda]).  Groups of [Task]s sent to [schedule] are guaranteed to be run in the order in which
+ * they were received, while the group's order itself is defined by a natural ordering of
+ * [Processor]s followed by grouped [Listener]s.
  */
-@Suppress("EXPERIMENTAL_API_USAGE")
+@Suppress("EXPERIMENTAL_API_USAGE", "MemberVisibilityCanBePrivate")
 class Scheduler(
-    private val time: Time,
     context: CoroutineContext,
-    private val agendaProcessingTimeoutMs: Long = DEFAULT_AGENDA_PROCESSING_TIMEOUT_MS,
-    private val scheduleRateHz: Int = DEFAULT_SCHEDULE_RATE_HZ
+    private val agendaProcessingTimeoutMs: Long = DEFAULT_AGENDA_PROCESSING_TIMEOUT_MS
 ) {
     private val log = TaggedLog { "Scheduler(${hashCode()})" }
     /* internal */
     val launches = atomic(0)
-    /* internal */
-    val loops = atomic(0)
     val scope = CoroutineScope(context)
 
-    private val isActive: Boolean
-        get() = !(processingJob?.isCompleted ?: true)
-    private var processingJob: Job? = null
-    private val isPaused = atomic(false)
-    private val agenda = atomic(Agenda())
+    private val agendaChannel = Channel<Agenda>(Channel.UNLIMITED)
+    private val pausedChannel = ConflatedBroadcastChannel(false)
+    private val idlenessChannel = ConflatedBroadcastChannel(true)
+
+    /**
+     * Flow of booleans indicating when the Scheduler has entered (`true`) or exited (`false`) and
+     * idle state.
+     */
+    val idlenessFlow: Flow<Boolean> = idlenessChannel.asFlow()
+
+    init {
+        // Consume the agenda channel:
+        // 1. Wait until the latest pause-value is false,
+        // 2. Notify the idleness channel that we're busy,
+        // 3. Do the work
+        // 4. Notify the idleness channel that we're not busy.
+        agendaChannel.consumeAsFlow()
+            .onEach { agenda ->
+                pausedChannel.asFlow().filter { !it }.first()
```

Comment:
Does this really make sense to use a flow? I'm wondering if it's overkill (channel + flow) vs a construct that just waits until an atomic boolean changes to false. It seems what you'd really want here is atomicBoolean.waitUntilFalse() if such a construct existed.

Reply:
I think this is ok for now, I'm just wondering how crazy we should go with flows, I really like them, but I worry about overhead if we go nuts. :)

Reply:
I agree... atomicBoolean.waitUntilFalse would be ideal here. I'll leave a comment/todo.